README
Documentation on http://wiki.github.com/augustl/time-dot-js

CHANGELOG
1.0.1
  * Added `week`, the current week of the year.

1.0
  * firstDayOfWeek is zero indexed for consistency with other languages. For example, in C, 0-6 implies sun-sat and 1-7 implies mon-sun.

1.0 RC 1
  * Creating first release candidate

test/tests.js
(function () {
    // Epoch in seconds, for test convenience. Comparing milliseconds sucks, because
    // time instances created runtime might differ with a few MS.
    Time.prototype.s = function(){
        return Math.floor(this.epoch() / 1000);
    }
    
    TestCase("Base test", {
        tearDown: function () {
            Time.firstDayOfWeek = 0;
        },
        
        "test basic accessors": function () {
            var now = new Time();
            var date = new Date();

            now.year(2008);
            assertEquals(now.year(), 2008);

            assertEquals(now.day(), date.getDate());
            now.day(5);
            assertEquals(now.day(), 5);
        },
        
        "test basic accessors": function () {
            var now = new Time();
            var date = new Date();

            now.year(2008);
            assertEquals(now.year(), 2008);

            assertEquals(now.day(), date.getDate());
            now.day(5);
            assertEquals(now.day(), 5);
        },

        "test week": function () {
            assertEquals(52, new Time(2010, 1, 1).week());
            assertEquals(1, new Time(2010, 1, 3).week());
            assertEquals(1, new Time(2010, 1, 4).week());
            assertEquals(2, new Time(2010, 1, 10).week());
            assertEquals(2, new Time(2010, 1, 11).week());

            Time.firstDayOfWeek = 1;
            assertEquals(53, new Time(2010, 1, 1).week());
            assertEquals(53, new Time(2010, 1, 3).week());
            assertEquals(1, new Time(2010, 1, 4).week());
            assertEquals(1, new Time(2010, 1, 10).week());
            assertEquals(2, new Time(2010, 1, 11).week());
        },
        
        "test weekday": function () {
            assertEquals(1, new Time(2008, 5, 11).weekday());
    	    assertEquals(7, new Time(2008, 5, 17).weekday());

    	    var time = new Time(2008, 5, 16);
    	    time.weekday(1);
    	    assertEquals(1, time.weekday());
    	    assertEquals(new Time(2008, 5, 11).s(), time.s());

    	    var time = new Time(2008, 5, 12);
    	    time.weekday(7);
    	    assertEquals(7, time.weekday());
    	    assertEquals(new Time(2008, 5, 17).s(), time.s());

    	    Time.firstDayOfWeek = 1

    	    assertEquals(6, new Time(2008, 5, 17).weekday())
    	    assertEquals(7, new Time(2008, 5, 18).weekday())

    	    var time = new Time(2008, 5, 17);
    	    time.weekday(1);
    	    assertEquals(1, time.weekday());
    	    assertEquals(new Time(2008, 5, 12).s(), time.s());

    	    var time = new Time(2008, 5, 13);
    	    time.weekday(7);
    	    assertEquals(7, time.weekday());
    	    assertEquals(new Time(2008, 5, 18).s(), time.s());

    	    Time.firstDayOfWeek = 4;
    	    assertEquals(1, new Time(2008, 5, 15).weekday());
    	    assertEquals(7, new Time(2008, 5, 21).weekday());
        },
        
        "test creating instance defaults to now": function () {
            var now = new Time();
            var date = new Date();

            assertEquals(now.year(), date.getFullYear());
            assertEquals(now.second(), date.getSeconds());
        },
        
        "test creating instance with string piggybacks to Date.parse": function () {
            var now = new Time("Jul 8, 2005");

            assertEquals(now.year(), 2005);
            assertEquals(now.month(), 7); // Should not be zero indexed
            assertEquals(now.day(), 8);
        },
        
        "test cloning": function () {
            var time = new Time(2008, 5);
            var clone = time.clone();

            assertEquals(clone.year(), 2008);
            assertEquals(clone.month(), 5);

            clone.year(1994);
            assertEquals(time.year(), 2008) // Shouldn't change
        },
        
        "test isLeapYear": function () {
            assertTrue(new Time(2008).isLeapYear());
            assertTrue(new Time(1804).isLeapYear());
            assertTrue(new Time(2292).isLeapYear());

            assertFalse(new Time(2009).isLeapYear());
        },
        
        "test days in month": function () {
            assertEquals(new Time(2008, 1).daysInMonth(), 31);
            assertTrue(new Time(2008, 2).daysInMonth() < 30); // leap years, hooray
            assertEquals(new Time(2008, 3).daysInMonth(), 31);
            assertEquals(new Time(2008, 4).daysInMonth(), 30);
            assertEquals(new Time(2008, 5).daysInMonth(), 31);
            assertEquals(new Time(2008, 6).daysInMonth(), 30);
            assertEquals(new Time(2008, 7).daysInMonth(), 31);
            assertEquals(new Time(2008, 8).daysInMonth(), 31);
            assertEquals(new Time(2008, 9).daysInMonth(), 30);
            assertEquals(new Time(2008, 10).daysInMonth(), 31);
            assertEquals(new Time(2008, 11).daysInMonth(), 30);
            assertEquals(new Time(2008, 12).daysInMonth(), 31);
        },
        
        "test weeks in month": function () {
            assertEquals(new Time(2008, 1).weeksInMonth(), 5);
    	    assertEquals(new Time(2008, 2).weeksInMonth(), 5);
    	    assertEquals(new Time(2008, 3).weeksInMonth(), 6);
    	    assertEquals(new Time(2008, 4).weeksInMonth(), 5);
    	    assertEquals(new Time(2008, 5).weeksInMonth(), 5);
    	    assertEquals(new Time(2008, 6).weeksInMonth(), 5);
    	    assertEquals(new Time(2008, 7).weeksInMonth(), 5);
    	    assertEquals(new Time(2008, 8).weeksInMonth(), 6);
    	    assertEquals(new Time(2008, 9).weeksInMonth(), 5);
    	    assertEquals(new Time(2008, 10).weeksInMonth(), 5);
    	    assertEquals(new Time(2008, 11).weeksInMonth(), 6);
    	    assertEquals(new Time(2008, 12).weeksInMonth(), 5);
        },
        
        "test week of current month": function () {
            assertEquals(1, new Time(2008, 11, 1).weekOfCurrentMonth());
            assertEquals(2, new Time(2008, 11, 2).weekOfCurrentMonth());
            assertEquals(2, new Time(2008, 11, 8).weekOfCurrentMonth());
            assertEquals(4, new Time(2008, 11, 18).weekOfCurrentMonth());
            assertEquals(6, new Time(2008, 11, 30).weekOfCurrentMonth());
            
            Time.firstDayOfWeek = 1;
            
            assertEquals(1, new Time(2008, 11, 1).weekOfCurrentMonth());
            assertEquals(1, new Time(2008, 11, 2).weekOfCurrentMonth());
            assertEquals(2, new Time(2008, 11, 8).weekOfCurrentMonth());
            assertEquals(4, new Time(2008, 11, 18).weekOfCurrentMonth());
            assertEquals(5, new Time(2008, 11, 30).weekOfCurrentMonth());
        },
        
        "test setting first day of week on instance": function () {
            assertEquals(1, new Time(2008, 4, 27).weekday())

            var t = new Time(2008, 4, 28);
            t.firstDayOfWeek = 1;
            assertEquals(1, t.weekday());

            // Instance scope overrides global scope.
            Time.firstDayOfWeek = 3
            assertEquals(1, t.weekday());
        }
    })
    
    TestCase("Stepping", {
        tearDown: function () {
            Time.firstDayOfWeek = 0;
        },
        
        "test beginning": function () {
            assertEquals(new Time(2008, 1, 0, 0, 0).s(),    new Time(2008, 5, 17, 15, 30).beginningOfYear().s());
            assertEquals(new Time(2008, 5, 0, 0, 0).s(),    new Time(2008, 5, 17, 15, 30).beginningOfMonth().s());
            assertEquals(new Time(2008, 5, 17, 0, 0).s(),   new Time(2008, 5, 17, 15, 30).beginningOfDay().s());
            assertEquals(new Time(2008, 5, 17, 15, 0).s(),  new Time(2008, 5, 17, 15, 30).beginningOfHour().s());
            assertEquals(new Time(2008, 5, 17, 15, 30).s(), new Time(2008, 5, 17, 15, 30).beginningOfMinute().s());

    	    assertEquals(new Time(2008, 5, 17).beginningOfWeek().s(), new Time(2008, 5, 11).s())

    	    Time.firstDayOfWeek = 1;

    	    assertEquals(new Time(2008, 5, 17).beginningOfWeek().s(), new Time(2008, 5, 12).s())
        },
        
        "test end": function () {
            assertEquals(new Time(2008, 12, 31, 23, 59, 59, 999).s(), new Time(2008, 5, 17, 15, 30).endOfYear().s());

            // January should have 31 days
            assertEquals(new Time(2008, 1, 31, 23, 59, 59, 999).s(), new Time(2008, 1, 23).endOfMonth().s());
            // April should have 30 days
            assertEquals(new Time(2008, 4, 30, 23, 59, 59, 999).s(), new Time(2008, 4, 15).endOfMonth().s());

            assertEquals(new Time(2008, 5, 17, 23, 59, 59, 999).s(), new Time(2008, 5, 17, 15, 30).endOfDay().s());
            assertEquals(new Time(2008, 5, 17, 15, 59, 59, 999).s(), new Time(2008, 5, 17, 15, 30).endOfHour().s());
            assertEquals(new Time(2008, 5, 17, 15, 30, 59, 999).s(), new Time(2008, 5, 17, 15, 30, 23).endOfMinute().s());

    	    assertEquals(new Time(2008, 5, 12).endOfWeek().s(), new Time(2008, 5, 17).endOfDay().s())

    	    Time.firstDayOfWeek = 1;

    	    assertEquals(new Time(2008, 5, 17).endOfWeek().s(), new Time(2008, 5, 18).endOfDay().s())
        },
        
        "test first day in calendar month": function () {
            assertEquals(new Time(2008, 1).firstDayInCalendarMonth().s(), new Time(2007, 12, 30).s());
    	    assertEquals(new Time(2008, 2).firstDayInCalendarMonth().s(), new Time(2008, 1, 27).s());
    	    assertEquals(new Time(2008, 3).firstDayInCalendarMonth().s(), new Time(2008, 2, 24).s());

    	    assertEquals(new Time(2008, 4).firstDayInCalendarMonth().s(), new Time(2008, 3, 30).s());

    	    Time.firstDayOfWeek = 1;
    	    assertEquals(new Time(2008, 4, 1).firstDayInCalendarMonth().s(), new Time(2008, 3, 31).s())
    	    assertEquals(new Time(2008, 3, 1).firstDayInCalendarMonth().s(), new Time(2008, 2, 25).s());
        },
        
        "test advance days": function () {
            assertEquals(new Time(2008, 5, 17).advanceDays(1).s(), new Time(2008, 5, 18).s());
    	    assertEquals(new Time(2008, 5, 17).advanceDays(-1).s(), new Time(2008, 5, 16).s());
    	    assertEquals(new Time(2008, 5, 17).advanceDays(-10).s(), new Time(2008, 5, 7).s());

    	    assertEquals(new Time(2008, 5, 17).advanceDays(20).s(), new Time(2008, 6, 6).s());
    	    assertEquals(new Time(2008, 5, 17).advanceDays(-20).s(), new Time(2008, 4, 27).s());

    	    // This will break with a simple epoch() implementation.
    	    assertEquals(new Time(2008, 4).advanceDays(-2).s(), new Time(2008, 3, 30).s());
    	    
            // When I'm running this test, it won't pass if the code isn't aware of daylight saving
            // swap on november 26th 2008.
            assertEquals(new Time(2008, 10, 26).advanceDays(1).s(), new Time(2008, 10, 27).s());
        },
        
        "test advance month": function () {
            assertEquals(new Time(2008, 5).advanceMonths(1).s(), new Time(2008, 6).s());
    	    assertEquals(new Time(2008, 1).advanceMonths(-1).s(), new Time(2007, 12).s());
    	    assertEquals(new Time(2008, 1, 31).advanceMonths(1).s(), new Time(2008, 2, 29).s()); // 2008 is a leap year
    	    assertEquals(new Time(2007, 12, 31).advanceMonths(2).s(), new Time(2008, 2, 29).s());
    	    assertEquals(new Time(2007, 1, 31).advanceMonths(1).s(), new Time(2007, 2, 28).s())
    	    assertEquals(new Time(2008, 1, 3).advanceMonths(-13).s(), new Time(2006, 12, 3).s())
        },
        
        "test next month": function () {
            assertEquals(new Time(2008, 6, 1).s(), new Time(2008, 5, 17).nextMonth().s());
            assertEquals(new Time(2009, 1, 1).s(), new Time(2008, 12, 24).nextMonth().s());
        },
        
        "test previous month": function () {
            assertEquals(new Time(2008, 4, 1).s(), new Time(2008, 5, 17).previousMonth().s());
            assertEquals(new Time(2007, 12, 1).s(), new Time(2008, 1, 5).previousMonth().s());
        }
    })
}());


time.js
(function(){
    var Time = function(yearOrString, month, day, hour, minute, second, millisecond){
        if (typeof(yearOrString) === "string") {
            this.date = new Date(Date.parse(yearOrString));
        } else {
            if (yearOrString) {
                this.date = new Date(yearOrString, (month - 1 || 0), (day || 1), (hour || 0), (minute || 0), (second || 0), (millisecond || 0));
            } else {
                this.date = new Date();
            }
        }
    };

    Time.firstDayOfWeek = 0;
    var DAYS_IN_MONTH = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
    var MILLISECONDS_IN_WEEK = 604800000.0;
    var MILLISECONDS_IN_DAY = 86400000;
    var MILLISECONDS_IN_HOUR = 3600000;
    var THIRTY_TWO_DAYS = 2764800000;

    Time.prototype = {
        // Month gets special treatment, to avoid zero indexing from Date.
        month: function(value) {
            if (value) {
                this.date.setMonth(value - 1);
            } else {
                return this.date.getMonth() + 1;
            }
        },

        // The week number of the year, iso8601.
        week: function (value) {
            if (value) {
                throw new Error("Setting week number not implemented yet.");
            } else {
                // With thanks to http://github.com/jquery/jquery-ui/blob/37e8dd605da5d99600c0/ui/jquery.ui.datepicker.js#L919
                var checker = this.clone();
                checker.day(checker.day() + 4 - checker.weekday());
                var epoch = checker.epoch();
                checker.beginningOfYear();
                return Math.floor(Math.round((epoch - checker.epoch()) / MILLISECONDS_IN_DAY) / 7) + 1;
            }
        },
        
        // There is no setDay(), implementing that here.
        weekday: function(value) {
            if (value) {
                this.advanceDays(value - this.weekday());
            } else {
                var a = (this.date.getDay() - (this.firstDayOfWeek || Time.firstDayOfWeek));
                var b = 7;
                var ringModulo = a - Math.floor(a / b) * b;
                return ringModulo + 1;
            }
        },
        
        /////////////////////////////////////
        // Utility
        
        clone: function(){
            var newTime = new Time();
            newTime.epoch(this.epoch());
            newTime.firstDayOfWeek = this.firstDayOfWeek;

            return newTime;
        },

        isLeapYear: function(){
            var year = this.year();
            return (year % 4 === 0) && (year % 100 !== 0) || (year % 400 === 0);
        },

        daysInMonth: function(){
            if (this.month() === 2 && this.isLeapYear()) {
                return 29;
            }

            return DAYS_IN_MONTH[this.month() - 1];
        },

        weeksInMonth: function(){
            var millisecondsInThisMonth = this.clone().endOfMonth().epoch() - this.clone().firstDayInCalendarMonth().epoch();
            return Math.ceil(millisecondsInThisMonth / MILLISECONDS_IN_WEEK);
        },

        // The current week's index in weeksInMonth
        weekOfCurrentMonth: function(){
            // NFI how this works. The tests pass, and brute force ftw.
            return Math.floor((this.day() + this.clone().day(1).weekday() + 5) / 7);
        },

        firstDayInCalendarMonth: function(){
            this.beginningOfMonth();
            this.weekday(1);
            return this;
        },

        /////////////////////////////////////
        // Stepping
        beginningOfYear: function(){
            this.month(1); this.beginningOfMonth();
            return this;
        },

        beginningOfMonth: function(){
            this.day(1); this.beginningOfDay();
            return this;
        },

        beginningOfWeek: function(){
            this.weekday(1); this.beginningOfDay();
            return this;
        },

        beginningOfDay: function(){
            this.hour(0); this.beginningOfHour();
            return this;
        },

        beginningOfHour: function(){
            this.minute(0); this.beginningOfMinute();
            return this;
        },

        beginningOfMinute: function(){
            this.second(0); this.millisecond(0);
            return this;
        },

        // ------------

        endOfYear: function(){
            this.month(12); this.endOfMonth();
            return this;
        },

        endOfMonth: function(){
            this.day(this.daysInMonth()); this.endOfDay();
            return this;
        },

        endOfWeek: function(){
            this.weekday(7); this.endOfDay();
            return this;
        },

        endOfDay: function(){
            this.hour(23); this.endOfHour();
            return this;
        },

        endOfHour: function(){
            this.minute(59); this.endOfMinute();
            return this;
        },

        endOfMinute: function(){
            this.second(59);
            return this;
        },

        // ------------

        nextMonth: function(){
            this.epoch(this.beginningOfMonth().epoch() + THIRTY_TWO_DAYS);
            this.beginningOfMonth();
            return this;
        },

        previousMonth: function(){
            this.epoch(this.beginningOfMonth().epoch() - 1);
            this.beginningOfMonth();
            return this;
        },

        // ------------

        advanceDays: function(days) {
            this.beginningOfDay();
            // Adding 1 hour, in case this is a day where daylight savings change
            this.epoch(this.epoch() + (MILLISECONDS_IN_DAY * days) + MILLISECONDS_IN_HOUR);
            this.beginningOfDay();

            return this;
        },

        advanceMonths: function(months) {
            var base = this.year() * 12 + (this.month() - 1) + months;
            var newYear = Math.floor(base / 12);
            var newMonth = (base % 12) + 1;

            // Setting the month to '2' on january 31th
            // gives us march 2nd. Circumventing this.
            var newTime = new Time(newYear, newMonth);
            var daysInNewTimeMonth = newTime.daysInMonth();

            if (this.day() > daysInNewTimeMonth) {
                this.year(newYear);
                this.day(1);
                this.month(newMonth);
                this.day(daysInNewTimeMonth);
            } else {
                this.year(newYear);
                this.month(newMonth);
            }

            return this;
        },


        // ------------

        toString: function() {
            return [this.year(), this.month(), this.day()].join(".") + " " + [this.hour(), this.minute(), this.second()].join(":");
        }
    };
    
    /////////////////////////////////////
    // The accessors. Uses the same function for both getting and
    // setting, e.g. year() to get and year(2005) to set.
    Time.accessor = function(name, dateFunctionName){
        Time.prototype[name] = function(value) {
            if (value !== undefined) {
                this.date["set" + dateFunctionName](value);
                return this;
            } else {
                return this.date["get" + dateFunctionName]();
            }
        };
    };
    
    Time.accessor("year", "FullYear");
    // month: see below
    Time.accessor("day", "Date");
    Time.accessor("hour", "Hours");
    Time.accessor("minute", "Minutes");
    Time.accessor("second", "Seconds");
    Time.accessor("millisecond", "Milliseconds");
    // weekday: See below
    Time.accessor("epoch", "Time");

    // Todo: Some kind of noConflict() thing, in case people have a 'Time' around already.
    window.Time = Time;
})();